## Load data

In [1]:
import pandas as pd

customers = pd.read_csv('../data/customer.csv')
products = pd.read_csv('../data/products.csv')
orders = pd.read_csv('../data/orders.csv')

## Inspect customers

In [2]:
customers.head()


,customer_id,gender,age,city,signup_date,loyalty_member
0,C0001,Male,19,Liverpool,2023-10-17,Yes
1,C0002,Male,65,Manchester,2022-08-17,Yes
2,C0003,Female,20,London,2023-08-28,Yes
3,C0004,Male,50,London,2022-08-12,Yes
4,C0005,Female,32,Nottingham,2023-10-28,No


In [3]:
# basic info
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   customer_id     1200 non-null   str  
 1   gender          1200 non-null   str  
 2   age             1200 non-null   int64
 3   city            1200 non-null   str  
 4   signup_date     1200 non-null   str  
 5   loyalty_member  1200 non-null   str  
dtypes: int64(1), str(5)
memory usage: 56.4 KB


In [4]:
# empty rows check
customers.isnull().sum()

customer_id       0
gender            0
age               0
city              0
signup_date       0
loyalty_member    0
dtype: int64

In [5]:
# check global duplicated rows
customers.duplicated().sum()

np.int64(0)

In [6]:
# check customers id duplicated
customers['customer_id'].duplicated().sum()

np.int64(0)

In [7]:
# gender values check
customers['gender'].value_counts()

gender
Male      613
Female    587
Name: count, dtype: int64

In [8]:
# check min & max customer's ages
customers['age'].describe()

count    1200.000000
mean       41.615833
std        13.719515
min        18.000000
25%        30.000000
50%        41.000000
75%        54.000000
max        65.000000
Name: age, dtype: float64

In [9]:
#check values in age row
customers['age'].value_counts().sort_index()

age
18    17
19    24
20    25
21    22
22    24
23    26
24    29
25    24
26    30
27    22
28    22
29    29
30    23
31    23
32    28
33    26
34    20
35    22
36    24
37    31
38    28
39    31
40    31
41    27
42    36
43    23
44    16
45    22
46    27
47    28
48    22
49    17
50    20
51    23
52    25
53    32
54    28
55    20
56    27
57    28
58    23
59    27
60    22
61    23
62    25
63    26
64    25
65    27
Name: count, dtype: int64

In [10]:
# count ages groups
customers['age_group'] = pd.cut(
    customers['age'],
    bins=[17, 25, 35, 45, 55, 65],
    labels=['18-25', '26-35', '36-45', '46-55', '56-65']
)
customers['age_group'].value_counts().sort_index()


age_group
18-25    191
26-35    245
36-45    269
46-55    242
56-65    253
Name: count, dtype: int64

In [11]:
# cities values check
customers['city'].value_counts().sort_index()

city
Birmingham    159
Bristol       156
Leeds         146
Liverpool     137
London        146
Manchester    133
Nottingham    160
Sheffield     163
Name: count, dtype: int64

In [12]:
# signup dates values check
customers['signup_date'].value_counts().sort_index()

signup_date
2022-01-01    1
2022-01-03    2
2022-01-04    2
2022-01-05    1
2022-01-06    1
             ..
2023-11-28    3
2023-11-29    1
2023-11-30    1
2023-12-01    2
2023-12-02    1
Name: count, Length: 574, dtype: int64

In [13]:
customers['signup_date'].info()

<class 'pandas.Series'>
RangeIndex: 1200 entries, 0 to 1199
Series name: signup_date
Non-Null Count  Dtype
--------------  -----
1200 non-null   str  
dtypes: str(1)
memory usage: 9.5 KB


## Inspect products

In [14]:
products.head()

,product_id,product_name,category,price
0,P001,Blender 1,Beauty,272.76
1,P002,Perfume 2,Home,132.09
2,P003,Water Bottle 3,Home,58.58
3,P004,Smart Watch 4,Home,220.82
4,P005,Perfume 5,Home,27.81


In [15]:
products.isnull().sum()

product_id      0
product_name    0
category        0
price           0
dtype: int64

In [16]:
products.info()

<class 'pandas.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    80 non-null     str    
 1   product_name  80 non-null     str    
 2   category      80 non-null     str    
 3   price         80 non-null     float64
dtypes: float64(1), str(3)
memory usage: 2.6 KB


In [17]:
products['product_id'].duplicated().sum()

np.int64(0)

In [18]:
products['product_name'].duplicated().sum()

np.int64(0)

In [19]:
products['product_name'].str.strip().str.lower().duplicated().sum()

np.int64(0)

In [20]:
products['product_name'].str.strip().eq('').sum()

np.int64(0)

In [21]:
products['product_name'].value_counts().head(20)

product_name
Blender 1               1
Perfume 2               1
Water Bottle 3          1
Smart Watch 4           1
Perfume 5               1
Vacuum Cleaner 6        1
Laptop Stand 7          1
Smart Watch 8           1
USB Cable 9             1
Yoga Mat 10             1
Yoga Mat 11             1
Wireless Mouse 12       1
Perfume 13              1
Jeans 14                1
Running Shoes 15        1
Hair Dryer 16           1
USB Cable 17            1
Bluetooth Speaker 18    1
Coffee Maker 19         1
Blender 20              1
Name: count, dtype: int64

In [22]:
products['category'].value_counts().sort_index()

category
Beauty         10
Clothing       15
Electronics    13
Home           21
Sports         21
Name: count, dtype: int64

In [23]:
# check uncorrected values in prices
products[pd.to_numeric(products['price'], errors='coerce').isnull()]

,product_id,product_name,category,price


In [24]:
# check min & max prices values
products['price'].describe()

count     80.000000
mean     140.082875
std       87.534130
min       11.070000
25%       58.922500
50%      126.115000
75%      214.132500
max      289.840000
Name: price, dtype: float64

## Inspect orders

In [25]:
orders.head()

,order_id,customer_id,product_id,order_date,quantity,payment_method
0,O00001,C0692,P006,2023-01-29,4,Card
1,O00002,C1072,P059,2023-04-19,5,Card
2,O00003,C1187,P030,2024-01-15,1,Card
3,O00004,C0290,P063,2024-01-12,1,Cash
4,O00005,C0094,P032,2024-04-09,2,Card


In [26]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   order_id        4000 non-null   str  
 1   customer_id     4000 non-null   str  
 2   product_id      4000 non-null   str  
 3   order_date      4000 non-null   str  
 4   quantity        4000 non-null   int64
 5   payment_method  4000 non-null   str  
dtypes: int64(1), str(5)
memory usage: 187.6 KB


In [27]:
orders.isnull().sum()

order_id          0
customer_id       0
product_id        0
order_date        0
quantity          0
payment_method    0
dtype: int64

In [28]:
orders['order_id'].duplicated().sum()

np.int64(0)

In [29]:
orders['payment_method'].value_counts()

payment_method
Cash      1362
Card      1345
Online    1293
Name: count, dtype: int64

## Data Cleaning
- Missing values: not found
- Duplicates: not found (including primary keys)
- Data types: correct
- Numeric columns: no invalid values
- Categorical columns: no obvious anomalies

Conclusion: no cleaning required.

## Load to SQLite


In [30]:
import sqlite3

# Create database
conn = sqlite3.connect('../database/retail.db')

# Load data into SQLite database
customers.to_sql('customers', conn, if_exists='replace', index=False)
products.to_sql('products', conn, if_exists='replace', index=False)
orders.to_sql('orders', conn, if_exists='replace', index=False)

# Check success
pd.read_sql("SELECT * FROM customers LIMIT 5", conn)



,customer_id,gender,age,city,signup_date,loyalty_member,age_group
0,C0001,Male,19,Liverpool,2023-10-17,Yes,18-25
1,C0002,Male,65,Manchester,2022-08-17,Yes,56-65
2,C0003,Female,20,London,2023-08-28,Yes,18-25
3,C0004,Male,50,London,2022-08-12,Yes,46-55
4,C0005,Female,32,Nottingham,2023-10-28,No,26-35


# Business Questions

## Total number of customers
Result: 1200 customers

In [31]:

pd.read_sql("SELECT COUNT(customer_id) as count_of_customers FROM customers", conn)

,count_of_customers
0,1200


## Total numbers of orders
Result: 4000 orders

In [32]:
pd.read_sql("SELECT COUNT(order_id) as count_of_orders FROM orders", conn)

,count_of_orders
0,4000


## Top cities by number of customers
Result: Sheffield, Nottingham, Birmingham, Bristol

In [67]:
pd.read_sql(
    """SELECT city, COUNT(customer_id)
    FROM customers
    GROUP BY city
    ORDER BY COUNT(customer_id) DESC
    LIMIT 4
            """, conn)

,city,COUNT(customer_id)
0,Sheffield,163
1,Nottingham,160
2,Birmingham,159
3,Bristol,156


## Total revenue
Result: 1644817.71

In [34]:
# JOIN product_id from different tables
pd.read_sql("""
SELECT SUM(products.price * orders.quantity) as total_revenue
FROM orders
JOIN
products ON orders.product_id = products.product_id""", conn).head()

,total_revenue
0,1644817.71


## Top 5 most purchased products
### Result:
1. Monitor 55 (201)
2. Wireless Mouse 60 (193)
3. Bluetooth Speaker 18 (191)
4. Water Bottle 71 (178)
5. USB Cable 65 (178)

In [68]:
pd.read_sql("""
SELECT SUM(orders.quantity),
    products.product_name
FROM orders
JOIN products ON orders.product_id = products.product_id
GROUP BY products.product_name
ORDER BY SUM(orders.quantity) DESC
""", conn).head()

,SUM(orders.quantity),product_name
0,201,Monitor 55
1,193,Wireless Mouse 60
2,191,Bluetooth Speaker 18
3,178,Water Bottle 71
4,178,USB Cable 65


## Top category by revenue
### Result: Home (473002.14)

In [38]:
pd.read_sql("""
SELECT products.category as ranking_category_by_revenue,
    SUM(orders.quantity * products.price) as total_revenue_by_category
FROM orders
JOIN products ON orders.product_id = products.product_id
GROUP BY products.category
ORDER BY SUM(orders.quantity * products.price) DESC
""", conn).head()

,ranking_category_by_revenue,total_revenue_by_category
0,Home,473002.14
1,Sports,403963.89
2,Clothing,319942.70
3,Beauty,239438.96
4,Electronics,208470.02


## Average quantity per order
### Result: 2.97

In [39]:
pd.read_sql("""
SELECT SUM(orders.quantity) * 1.0 / COUNT(DISTINCT orders.order_id) as average_order_quantity
FROM orders
""", conn).head()

,average_order_quantity
0,2.97475


In [40]:
pd.read_sql("""SELECT * FROM customers LIMIT 5""", conn)

,customer_id,gender,age,city,signup_date,loyalty_member,age_group
0,C0001,Male,19,Liverpool,2023-10-17,Yes,18-25
1,C0002,Male,65,Manchester,2022-08-17,Yes,56-65
2,C0003,Female,20,London,2023-08-28,Yes,18-25
3,C0004,Male,50,London,2022-08-12,Yes,46-55
4,C0005,Female,32,Nottingham,2023-10-28,No,26-35


In [41]:
pd.read_sql("""SELECT * FROM products LIMIT 5""", conn)

,product_id,product_name,category,price
0,P001,Blender 1,Beauty,272.76
1,P002,Perfume 2,Home,132.09
2,P003,Water Bottle 3,Home,58.58
3,P004,Smart Watch 4,Home,220.82
4,P005,Perfume 5,Home,27.81


In [42]:
pd.read_sql("""SELECT * FROM orders LIMIT 5""", conn)

,order_id,customer_id,product_id,order_date,quantity,payment_method
0,O00001,C0692,P006,2023-01-29,4,Card
1,O00002,C1072,P059,2023-04-19,5,Card
2,O00003,C1187,P030,2024-01-15,1,Card
3,O00004,C0290,P063,2024-01-12,1,Cash
4,O00005,C0094,P032,2024-04-09,2,Card


## Top cities by revenue
### Result: Sheffield (233607.88)

In [43]:
pd.read_sql("""
SELECT customers.city,
SUM(products.price * orders.quantity) as total_revenue_by_city
FROM orders
JOIN products ON orders.product_id = products.product_id
JOIN customers ON orders.customer_id = customers.customer_id
GROUP BY customers.city
ORDER BY SUM(products.price * orders.quantity) DESC
""", conn)

,city,total_revenue_by_city
0,Sheffield,233607.88
1,Birmingham,224892.95
2,Nottingham,213151.66
3,Bristol,202317.33
4,Liverpool,197613.63
5,Leeds,194316.01
6,London,194041.93
7,Manchester,184876.32


## Top 10 customers by spending

In [69]:
pd.read_sql("""
SELECT orders.customer_id AS customer_id,
SUM(orders.quantity * products.price) as total_spending
FROM orders
JOIN products ON orders.product_id = products.product_id
GROUP BY orders.customer_id
ORDER BY SUM(orders.quantity * products.price) DESC
""", conn).head(10)

,customer_id,total_spending
0,C0299,5124.21
1,C0094,5078.51
2,C0588,4874.30
3,C0826,4685.22
4,C0334,4515.14
5,C0291,4508.31
6,C0765,4505.37
7,C0157,4459.55
8,C1167,4425.53
9,C0674,4356.89


## Best month by revenue

### Result: top month by revenue is January of 2024 (127940.85)

In [60]:
pd.read_sql("""
SELECT SUM(orders.quantity * products.price) as monthly_revenue,
strftime('%Y-%m', order_date) as month
FROM orders
JOIN products ON orders.product_id = products.product_id
GROUP BY strftime('%Y-%m', order_date)
ORDER BY monthly_revenue DESC
""",conn)

,monthly_revenue,month
0,127940.85,2024-01
1,108265.02,2023-01
2,108166.21,2024-02
3,107246.94,2023-06
4,103306.51,2023-03
5,101482.34,2023-02
6,100780.97,2023-07
7,100330.40,2024-04
8,97081.04,2023-08
9,97011.30,2024-03


## Most common payment method
### Result: Cash (1362)

In [65]:
pd.read_sql("""
SELECT COUNT(orders.payment_method) as payment_count,
    orders.payment_method
FROM orders
GROUP BY orders.payment_method
ORDER BY payment_count DESC
""", conn)

,payment_count,payment_method
0,1362,Cash
1,1345,Card
2,1293,Online


## Key insights

- Total revenue: 1,644,817.71
- Highest revenue category: Home
- Highest revenue city: Sheffield
- Best month by revenue: 2024-01
- Most common payment method: Cash
- Top product by quantity sold: Monitor 55